### PR TITLE
Simplify auth flow types

### PR DESCRIPTION
**Pull request by nmattia**
This deduplicates the LoginSuccess & LoginFlow(Success,Result,Error) types and brings all error clarification to the top level.

Before this, several types (esp. `LoginSuccess` & `LoginFlowSuccess`) were used for similar purposes, with a translation layer (`apiResultToLoginFlowResult`) where necessary.

Moreover, a hierarchy of types existed (`ApiResult`, `RegisterResult`, ...) which obfuscated the actual subtypes and caused more error handling than actually necessary (by bundling types and surfacing errors that could never be triggered).

The changes implemented here do away with the "FlowResult" types (`LoginFlowResult`, `flowResult.ts`) and removes the artificial hierarchy between types (though introducing a new `FlowError` type representing a group of error types commonly found together).

Moreover the translation between error type and humand readable errors is left to the top-level caller, meaning error types stay in a more precise (i.e. non-string) format for longer.

These changes lead to less error handling and more precise error reporting, as well as fewer types to handle and translate.

### CODE REVIEW

File: src/frontend/src/components/authenticateBox.ts [hunk 1]
```diff
@@ -256,27 +280,83 @@ export const authenticateBoxFlow = async <T, I>({
...
+
+// Makes the error human readable
+const clarifyError: {
+  [K in FlowError["kind"]]: (err: AssociatedError<K>) => {
+    title: string;
+    message: string;
+    detail?: string;
+  };
+} = {
+  authFail: (err) => ({
```
Comment (nmattia) [inline src/frontend/src/components/authenticateBox.ts:309]: All these messages were copied from now-gone `flowResult.ts`

File: src/frontend/src/flows/recovery/recoverWith/phrase.ts [hunk 2]
```diff
@@ -312,36 +303,40 @@ export const wordTemplate = ({
...
+        const result = await withLoader(() =>
+          connection.fromSeedPhrase(userNumber, words.join(" "))
         );
-        return result;
+        if (result.kind !== "loginSuccess") {
+          const msg = {
+            noSeedPhrase:
+              "No recovery phrase: there is no recovery phrase associated with this identity. Cancel and recover another way.",
+            seedPhraseFail:
+              "Invalid Seed Phrase: Failed to authenticate using this seed phrase. Did you enter it correctly?",
```
Comment (nmattia) [inline src/frontend/src/flows/recovery/recoverWith/phrase.ts:332]: Taken from `flowResult.ts` where they could have actually never been triggered (there errors were not allowed to bubble up) and added here

File: src/frontend/src/components/authenticateBox.ts [hunk 3]
```diff
@@ -256,27 +280,83 @@ export const authenticateBoxFlow = async <T, I>({
...
+  | WebAuthnFailed
+  | UnknownUser
+  | ApiError
+  | RegisterNoSpace;
+
+// Type machinery to allow translating error code (e.g. kind: "authFail") using the actual data
+// contained in the error objects (e.g. authFail: (authFail: AuthFail) => ...).
+type ErrorKind = FlowError["kind"]; // all known error tags for FlowError
+type KindToError = { [P in ErrorKind]: Omit<FlowError & { kind: P }, "kind"> }; // Look up from err. kind to object
+type AssociatedError<K extends ErrorKind> = {
```
Comment (nmattia) [inline src/frontend/src/components/authenticateBox.ts:297]: I wonder if we can squash those two

File: src/frontend/src/components/authenticateBox.ts [hunk 4]
```diff
@@ -256,27 +280,83 @@ export const authenticateBoxFlow = async <T, I>({
...
+  | BadPin
+  | BadChallenge
+  | WebAuthnFailed
+  | UnknownUser
+  | ApiError
+  | RegisterNoSpace;
+
+// Type machinery to allow translating error code (e.g. kind: "authFail") using the actual data
+// contained in the error objects (e.g. authFail: (authFail: AuthFail) => ...).
+type ErrorKind = FlowError["kind"]; // all known error tags for FlowError
```
Comment (lmuntaner) [inline src/frontend/src/components/authenticateBox.ts:295]: Do they all the FlowError have a `"kind"` property?

File: src/frontend/src/components/authenticateBox.ts [hunk 5]
```diff
@@ -256,27 +280,83 @@ export const authenticateBoxFlow = async <T, I>({
...
+type KindToError = { [P in ErrorKind]: Omit<FlowError & { kind: P }, "kind"> }; // Look up from err. kind to object
```
Comment (lmuntaner) [inline src/frontend/src/components/authenticateBox.ts:296]: Could you add a comment of what this translates to with an example?

I had to ask chat gpt to translate it for me 😅

File: src/frontend/src/components/authenticateBox.ts [hunk 6]
```diff
@@ -256,27 +280,83 @@ export const authenticateBoxFlow = async <T, I>({
...
+type AssociatedError<K extends ErrorKind> = {
+  [P in K]: { kind: P } & KindToError[P];
+}[K]; // Helper for adding the kind to the object (otherwise TS gets a bit lost)
```
Comment (lmuntaner) [inline src/frontend/src/components/authenticateBox.ts:299]: I'm a bit lost also 😂

File: src/frontend/src/components/authenticateBox.ts (hunk 4 above)
Comment (nmattia) [reply src/frontend/src/components/authenticateBox.ts:295]: Yes! Otherwise this would be a type error I think

**Review by lmuntaner — APPROVED** [review]
Rubber stamp!

**Pull request merged**